feat(gatsby, gatsby-plugin-utils): add image cdn source urls to redux

veryspry opened the pull request:
<!--
  Have any questions? Check out the contributing docs at https://gatsby.dev/contribute, or
  ask in this Pull Request and a Gatsby maintainer will be happy to help :)

  For any major changes, please first open a bug report (if it's a bug) or a feature request.
-->

## Description

This gives us a way to count the number of unique Image CDN source urls that are processed during a gatsby build. The one downside being that we cannot include images from pages that use SSR or DSG since those are not processed during the initial build.

- Source URL's are added to a new object in the store called `telemetry` and stored in the `gatsbyImageSourceUrls` JS Set inside of that. (i.e. source URL's are added to `telemetry.gatsbyImageSourceUrls`)
- I chose to use a Set for this since we only want to store unique values which Set does this by default
- Adds an action / reducer case to clear this set so that the Inc Build CLI can clear this piece of Gatsby state after it reads it and reports it back to BigQuery / Gatsby Cloud
- Source URL's are added only if `process.env.GATSBY_SHOULD_TRACK_IMAGE_CDN_URLS` is set to `"1"` or `"true"`. Gatsby Cloud will take care of clearing out the source url's set so this env var helps to prevent any memory leak type behavior in local dev or deployment environments other than Gatsby Cloud where these urls would not get automatically cleaned

TylerBarnes reviewed (changes requested):
I think as part of this PR we should send the telemtry via IPC to Cloud

veryspry commented:
> I think as part of this PR we should send the telemtry via IPC to Cloud

We send telemetry via IPC within our cloud runner CLI so we'll need an additional PR there rather than here. That should be incoming early next week!

That said, this shouldn't hurt anything to be merged before that other work is done since it is ultimately just some extra data in memory :slightly_smiling_face:

TylerBarnes commented:
We need to send from oss before cloud can recieve it. There's a part in oss that's sending data already so we should add this data there (not sure where it is)

j0sh77 commented on packages/gatsby/index.d.ts:
```diff
@@ -1263,6 +1263,12 @@ export interface Actions {
     plugin?: ActionPlugin
   ): Promise<unknown>
 
+  /** @todo add docs reference */
```
Our `index.d.ts` file is manual right now. Seems good to me!

veryspry replied:
Nice! Just updated it :raised_hands:

j0sh77 commented:
This looks good to me! More data is awesome. 

```
TypeError: actions.addGatsbyImageSourceUrl is not a function
```

There are some test failures that are stemming from an (assumed) mocked `actions` object.

veryspry commented:
> This looks good to me! More data is awesome.
> 
> ```
> TypeError: actions.addGatsbyImageSourceUrl is not a function
> ```
> 
> There are some test failures that are stemming from an (assumed) mocked `actions` object.

Good catch @j0sh77! I updated the tests and also added a new test to validate that `gatsbyImageReslover` is in fact adding new image source urls to the redux store

veryspry commented:
Published canary with these changes:

```
gatsby@4.13.0-alpha-image-cdn-telemetry.23+aa922053d3
```

wardpeet commented on packages/gatsby/src/redux/index.ts:
```diff
@@ -37,6 +37,12 @@ export const readState = (): IGatsbyState => {
       })
     }
 
+    if (!state.telemetry) {
+      state.telemetry = {
+        gatsbyImageSourceUrls: new Set<string>(),
+      }
+    }
```
This shouldn't be necessary. Did you have an example where this fails without it?

veryspry replied:
My thought was that I didn't want to overwrite anything if it already existed, but since we don't persist `telemetry` to disk, than this would always be empty. Is that a correct assumption?

wardpeet replied:
Your reducer shoulds et the initial state, nothing else

veryspry replied:
Got it :+1:

wardpeet commented on packages/gatsby/package.json:
```diff
@@ -87,7 +87,7 @@
     "gatsby-plugin-page-creator": "^4.13.0-next.0",
     "gatsby-plugin-typescript": "^4.13.0-next.0",
-    "gatsby-plugin-utils": "^3.7.0-next.0",
+    "gatsby-plugin-utils": "3.7.0-alpha-image-cdn-telemetry.29",
```
Revert
```suggestion
    "gatsby-plugin-utils": "^3.7.0-next.0",
```

pieh commented on packages/gatsby/src/redux/reducers/telemetry.ts:
```diff
@@ -0,0 +1,46 @@
+  switch (action.type) {
+    case `PROCESS_GATSBY_IMAGE_SOURCE_URL`: {
+      const { sourceUrl } = action.payload
+      const nextState = new Set(state.gatsbyImageSourceUrls)
```
Similar thing for `MERGE_WORKER_QUERY_STATE` action

veryspry replied:
Got it! Updated sot that all these actions modify the existing state object instead of returning a new one:

- `PROCESS_GATSBY_IMAGE_SOURCE_URL`
- `CLEAR_GATSBY_IMAGE_SOURCE_URL`
- `MERGE_WORKER_QUERY_STATE`

TylerBarnes reviewed (commented):
Looking good! 🙌 💪

DanielSLew commented on packages/gatsby/src/redux/reducers/telemetry.ts:
```diff
@@ -0,0 +1,42 @@
+  action: ActionsUnion
+): IGatsbyState["telemetry"] => {
+  switch (action.type) {
+    case `PROCESS_GATSBY_IMAGE_SOURCE_URL`: {
```
I'm not seeing where this action or the one below would ever get dispatched, is that in a followup PR?

DanielSLew replied:
oh wait I see the add image action being dispatched, but I don't see the `clearGatsbyImageSourceUrls` action ever used.

veryspry replied:
yup, exactly! :slightly_smiling_face:

DanielSLew commented:
I have a general question, why did we opt to do this in the framework at all and not just handle this all in the inc-build-cli? We loop through all images already in the inc-build-cli so we could create a unique set there and write it to the disk in the same way between builds. This way it wouldn't be gatsby version dependent and we don't need to add anything to the framework that doesn't necessarily concern the framework and it would be more straightforward.

I would imagine the images that we are adding to the sourceUrls have some property on them that we could use in the inc-build-cli to say "add this image to source urls"/"don't add this image to source urls"

TylerBarnes commented on packages/gatsby-plugin-utils/src/polyfill-remote-file/__tests__/gatsby-image-resolver.ts:
```diff
@@ -726,4 +729,30 @@ describe(`gatsbyImageData`, () => {
+
+    process.env.GATSBY_SHOULD_TRACK_IMAGE_CDN_URLS = null
+
+    expect(actions.addGatsbyImageSourceUrl.mock.calls[0][0]).toEqual(
```
Can we use `.toHaveBeCalledWith()` instead of accessing the calls directly?

veryspry replied:
totally - will fix this up!

TylerBarnes commented on packages/gatsby/src/redux/actions/public.js:
```diff
@@ -1252,6 +1252,19 @@ actions.createJobV2 = (job: JobV2, plugin: Plugin) => (dispatch, getState) => {
   return createJobV2FromInternalJob(internalJob)(dispatch, getState)
 }
 
+actions.addGatsbyImageSourceUrl = (sourceUrl: string) => {
```
I don't think this needs to be a public action. We only need this internally right?

veryspry replied:
This one unfortunately needs to be public since only public actions get passed to `gatsbyImageResolver`

TylerBarnes replied:
Makes sense 👍

TylerBarnes commented on packages/gatsby/src/redux/actions/public.js:
```diff
@@ -1252,6 +1252,19 @@ actions.createJobV2 = (job: JobV2, plugin: Plugin) => (dispatch, getState) => {
+  }
+}
+
+actions.clearGatsbyImageSourceUrls = () => {
```
same with this one

veryspry replied:
This one can probably be private but made sense to organize it with "public" as well since `addGatsbyImageSourceUrl` [has to be public](https://github.com/gatsbyjs/gatsby/pull/35427#discussion_r878074878)

veryspry replied:
Done

TylerBarnes commented on packages/gatsby/src/redux/reducers/queries.ts:
```diff
@@ -225,6 +225,11 @@ export function queriesReducer(
     }
     case `MERGE_WORKER_QUERY_STATE`: {
+      // This action may be dispatched in cases where queries might not be included in the merge data
+      if (!action.payload.queryStateChunk) {
```
What made this start happening?

veryspry replied:
I didn't run into this happening exactly, but because `queryStateChunk` is added conditionally in the worker pool manager code, I wanted to make sure we didn't break things if it wasn't present for some reason

https://github.com/gatsbyjs/gatsby/pull/35427/files#diff-39c5caa1451bc42018f6769f9a1d9eb8877a213fa340c7bf23404b0f3cdd606bR130

veryspry replied:
I added a check for this in the `telemetry` reducer as well where we handle the same action: https://github.com/gatsbyjs/gatsby/pull/35427/files#diff-42f89f373af246dfb622f3f0824095d5840a7ce40f39b9f96b224aae8470899eR24-R27

TylerBarnes replied:
Ah got it, thanks for the explanation

pieh commented on packages/gatsby-plugin-utils/src/polyfill-remote-file/graphql/gatsby-image-resolver.ts:
```diff
@@ -241,6 +241,12 @@ export async function gatsbyImageResolver(
+  if (
+    [`true`, `1`].includes(process.env.GATSBY_SHOULD_TRACK_IMAGE_CDN_URLS || ``)
+  ) {
+    actions.addGatsbyImageSourceUrl(source.url)
```
Please do check if `addGatsbyImageSourceUrl` is available before trying to use it. Otherwise we might see crashes when older gatsby version is used with newer `gatsby-plugin-utils`.

veryspry replied:
Good idea - fixed!

pieh commented on packages/gatsby-plugin-utils/src/polyfill-remote-file/graphql/gatsby-image-resolver.ts:
```diff
@@ -241,6 +241,12 @@ export async function gatsbyImageResolver(
   }
 
+  if (
+    [`true`, `1`].includes(process.env.GATSBY_SHOULD_TRACK_IMAGE_CDN_URLS || ``)
```
This should be memoized / calculated once. `process.env` is slow and doing this on each image will add that penalty to all "images" ( https://github.com/nodejs/node/issues/3104 )

veryspry replied:
Fixed

veryspry commented:
Need to update the tests to have the env var `process.env.GATSBY_SHOULD_TRACK_IMAGE_CDN_URLS` set in some of them. That change is incoming
